pkg/controllers/tlspolicy/tlspolicy_certmanager_certificates.go

```diff
@@ -25,70 +25,88 @@ import (
 func (r *TLSPolicyReconciler) reconcileCertificates(ctx context.Context, tlsPolicy *v1alpha1.TLSPolicy, gwDiffObj *reconcilers.GatewayDiff) error {
 	log := crlog.FromContext(ctx)
 
+	log.V(3).Info("reconciling certificates")
 	for _, gw := range gwDiffObj.GatewaysWithInvalidPolicyRef {
 		log.V(1).Info("reconcileCertificates: gateway with invalid policy ref", "key", gw.Key())
 		if err := r.deleteGatewayCertificates(ctx, gw.Gateway, tlsPolicy); err != nil {
-			return err
+			return fmt.Errorf("error deleting certificates for gw %v: %w", gw.Gateway.Name, err)
 		}
 	}
 
 	// Reconcile Certificates for each gateway directly referred by the policy (existing and new)
 	for _, gw := range append(gwDiffObj.GatewaysWithValidPolicyRef, gwDiffObj.GatewaysMissingPolicyRef...) {
-		log.V(1).Info("reconcileCertificates: gateway with valid and missing policy ref", "key", gw.Key())
-		if err := r.reconcileGatewayCertificates(ctx, gw.Gateway, tlsPolicy); err != nil {
-			return err
+		log.V(1).Info("reconcileCertificates: gateway with valid or missing policy ref", "key", gw.Key())
+		expectedCertificates := r.expectedCertificatesForGateway(ctx, gw.Gateway, tlsPolicy)
+		if err := r.createOrUpdateGatewayCertificates(ctx, expectedCertificates); err != nil {
+			return fmt.Errorf("error creating and updating expected certificates for gateway %v: %w", gw.Gateway.Name, err)
+		}
+		if err := r.deleteUnexpectedCertificates(ctx, expectedCertificates, gw.Gateway, tlsPolicy); err != nil {
+			return fmt.Errorf("error removing unexpected certificate for gateway %v: %w", gw.Gateway.Name, err)
 		}
 	}
```

Comment:
Personally i kind of preferred the pattern this (and DNSRecord) was using of having a reconcileGateway\<Resource\> method that deals with reconciling resources for a given Gateway instead of having it broken up inside the main reconcile\<Resource\> method.

- reconcile\<Resource\> - Reconcile resources(create/update/delete) for the current policy
- reconcileGateway\<Resource\> - Reconcile resources(create/update/delete) for the current policy and gateway
- delete\<Resource\> - Delete resources for the current policy
- deleteGateway\<Resource\> - Delete resources for the current policy and gateway

I'd say if we are trying to make these controllers consistent (looks like you have copied the pattern from health checks) we should actually update the DNSRecord one to deal with unexpected gateway resources in the same way as the other two.

Since none of that is in the scope of this bug fix i'd say we can come back to this later.